pkg/installmanager/installmanager.go

```diff
@@ -351,6 +351,12 @@ func (m *InstallManager) Run() error {
 		m.log.WithError(err).Error("error while trying to preemptively clean up")
 		return err
 	}
+	// Fail the cluster provision if infraID is set. This can occur if previous install
+	// pod for current cluster provision is unexpectedly killed or evicted.
+	if provision.Spec.InfraID != nil {
```

Comment:
`cleanupFailedInstall` (invoked on L350) already has a code path that catches `provision.Spec.InfraID != nil` and cleans up the provision with that InfraID. So this code change says, "and also fail," right?

I don't have a better suggestion, but that was pretty confusing to figure out. I'll take another look next week.

Reply:
I had to stare at this for a moment as well. If I read things correctly, prevInfraID is set so that a subsequent provision attempt cleans up a provision that "cleanly" failed (ie provision Pod that didn't suddenly stop/get deleted/was rescheduled to a new node). This is a *good thing*. The previous code also catches the case where an infraID was set, and cleans things up since infraID != nil  is an unexpected situation at this point in the codeflow.

Then this new code is handling the "why on earth is infraID set at this point". The previous code cleaned things up, and this new if clause is making sure we don't bother going forward with a provision, and marks this provision attempt as a "we've recovered from a strange state, and will mark this attempt as an error so that on next attempt things will work out as expected".

With this if clause, we don't unexpectedly error later when we try to update the generated infraID. We mark it as a clean "this install failed" attempt, and a subsequent provision attempt will the the prevInfraID fields, the code will clean up (a no-op since we've just cleaned up in this current provision attempt).

Assuming I have explained this accurately, can we spread some comments around to explain why infraID != nil is unexpected at this point, and how we handle it so that resources don't get leaked, and future provision attempts have a chance at success?

I can't see how infraID != nil AND prevInfraID != nil at the same time, but perhaps a sanity check and a loud error message wouldn't hurt since this code is relying on some subtle interactions and sequences of events.

Reply:
@joelddiaz Everything you have said here is correct. I will go ahead and add some more details in the comments.

Reply:
> I can't see how infraID != nil AND prevInfraID != nil at the same time

Consider the following case
- cluster provision 1 fails due to a valid error. May be a throttling issue
- controller creates new provision 2 and transfers infraID from previous attempt to field prevInfraID.
- provision 2 install pod 1 gets killed/evicted after updating infraID
- provision 2 install pod 2 will see both infraID != nil and prevInfraID != nil

possible?

Reply:
We might have to run 2 separate cleanup runs here